webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart

```diff
@@ -9,81 +22,184 @@ class StandardMessagesManager {
   factory StandardMessagesManager() => _singleton;
 
   int getNextCategoryIndex() {
-    var lastIndex = categories.values.fold(0, (previousValue, category) => previousValue > category.categoryIndex ? previousValue : category.categoryIndex);
+    var lastIndex = localCategories.values.fold(0, (previousValue, category) => previousValue > category.categoryIndex ? previousValue : category.categoryIndex);
     return lastIndex + 1;
   }
 
   int getNextGroupIndexInCategory(String categoryId) {
-    var lastIndex = categories[categoryId].groups.values.fold(0, (previousValue, group) => previousValue > group.groupIndexInCategory ? previousValue : group.groupIndexInCategory);
+    var lastIndex = localCategories[categoryId].groups.values.fold(0, (previousValue, group) => previousValue > group.groupIndexInCategory ? previousValue : group.groupIndexInCategory);
     return lastIndex + 1;
   }
 
   int getNextMessageIndexInGroup(String categoryId, String groupId) {
-    var standardMessagesInGroup = categories[categoryId].groups[groupId].messages.values;
+    var standardMessagesInGroup = localCategories[categoryId].groups[groupId].messages.values;
     var lastIndexInGroup = standardMessagesInGroup.fold(0, (previousValue, r) => previousValue > r.indexInGroup ? previousValue : r.indexInGroup);
     return lastIndexInGroup + 1;
   }
 
-  Map<String, MessageCategory> categories = {};
+  Map<String, MessageCategory> storageCategories = {};
+  Map<String, MessageCategory> localCategories = {};
+
+  MessagesDiffData get diffData {
+    Set<String> unsavedCategoryIds = {};
+    Set<String> unsavedGroupIds = {};
+    Set<String> unsavedMessageIds = {};
+
+    List<model.SuggestedReply> editedMessages = [];
+    List<model.SuggestedReply> deletedMessages = [];
+
+    Map<String, model.SuggestedReply> storageMessages = {};
+    Map<String, model.SuggestedReply> localMessages = {};
+
+    for (var category in localCategories.values) {
+      category.messages.forEach((message) {
+        localMessages[message.suggestedReplyId] = message;
+      });
+    }
+
+    for (var category in storageCategories.values) {
+      category.messages.forEach((message) {
+        storageMessages[message.suggestedReplyId] = message;
+      });
+    }
+
+    Set<String> allMessageKeys = Set()..addAll(localMessages.keys)..addAll(storageMessages.keys);
```

Comment:
nit, there's a nice shortcut for making a list or set out of multiple lists/sets by prefixing them with `...` (you can also add individual items at the same time if needed, but it's not the case here)
```suggestion
    Set<String> allMessageKeys = {...localMessages.keys, ...storageMessages.keys};
```